### DISCUSSION TITLE
Configuring Pyright to Locate venv in a Microservices Project

### DISCUSSION THREAD

**Discussion by 19104146** [zed-industries/zed#17868] [Feature Requests]
Hello, everyone. Apologies if this sounds basic—I'm new to Python development. I'm working with microservices and am having trouble configuring Pyright. Pyright cannot locate the `venv` folder for each microservice in the root directory. I want Pyright to use the nearest `venv` or the one I'm using in the terminal for each file.

Currently, my workaround is to create a separate `pyproject.toml` for each microservice folder (e.g., `product_service/`) and open the microservice directory in my editor (Zed) instead of the root directory. This setup allows Zed to access the `pyproject.toml` configuration for each microservice. However, I'd prefer to maintain a single `pyproject.toml` for the entire project.

Is there a way to configure Pyright to correctly locate the `.venv` folder so I can use a single `pyproject.toml` for the whole project instead of having one per microservice?

Here's the project structure for reference:

```
microservice_example_app
├── customer_service
│   ├── .venv
│   ├── Dockerfile
│   ├── main.py
│   ├── pyproject.toml
│   └── requirements.txt
├── product_service
│   ├── .venv
│   ├── Dockerfile
│   ├── main.py
│   ├── pyproject.toml
│   └── requirements.txt
├── docker-compose.yml
├── .gitignore
├── pyproject.toml
└── README.md
```

**Comment by omarjatoi** [comment]
You can use [execution environment options](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#execution-environment-options) in your main Pyright configuration (`pyproject.toml`) and specify one for each micro-service. Let me know if that helps (I haven't tried this myself in Zed, but it should work _in theory_).

**Reply by 19104146** [reply]
Hello, thank you for your reply. However, my problem remains the same—Pyright still can't locate the proper venv.

Here is the configuration I used:

```toml
# pyproject.toml
[tool.pyright]
executionEnvironments = [
  { root = "./customer_service" },
  { root = "./product_service" },
]
```

**Reply by omarjatoi** [reply]
And there is still a `./customer_service/pyproject.toml` that specifies a path to the venv? I realize you mentioned wanting a _single_ `pyproject.toml` file at the top level, but my thought is if you can combine using execution environments **and** individual `pyproject.toml` files and be able to work on subdirectories/services without having to open the text editor at different root directories.

**Reply by 19104146** [reply]
Yes, I've tried both using a single root `pyproject.toml` and keeping individual `pyproject.toml` files with changes to the root `pyproject.toml`. Unfortunately, neither approach works.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
